snapcraft/cli/_errors.py

```diff
@@ -78,6 +76,128 @@
 logger = logging.getLogger(__name__)
 
 
+def _is_connected_to_tty() -> bool:
+    # Used by inner instance, SNAPCRAFT_HAS_TTY is set by outer instance.
+    if os.getenv("SNAPCRAFT_BUILD_ENVIRONMENT") == "managed-host":
+        return distutils.util.strtobool(os.getenv("SNAPCRAFT_HAS_TTY", "n")) == 1
+
+    return sys.stdout.isatty()
+
+
+def _is_reportable_error(exc_info) -> bool:
+    # Report non-snapcraft errors.
+    if not issubclass(exc_info[0], errors.SnapcraftError):
+        return True
+
+    # Report snapcraft 'reportable' errors.
+    if issubclass(exc_info[0], errors.SnapcraftReportableError):
+        return True
+
+    return False
+
+
+def _is_printable_traceback(exc_info, debug) -> bool:
+    # Always print with debug.
+    if debug:
+        return True
+
+    # If it is not a reportable error, then it is not a printable traceback.
+    if not _is_reportable_error(exc_info):
+        return False
+
+    # Print if not connected to a tty.
+    if not _is_connected_to_tty():
+        return True
+
+    # Print if not using snap.
+    if not snapcraft.internal.common.is_snap():
+        return True
+
+    return False
+
+
+def _handle_sentry_submission(exc_info) -> None:
+    # Only attempt to submit reportable errors.
+    if not _is_reportable_error(exc_info):
+        return
+
+    # Only attempt if running as snap (with Raven requirement).
+    if not snapcraft.internal.common.is_snap():
+        # Suggest manual reporting instead.
+        click.echo(_MSG_MANUALLY_REPORT)
+        return
+
+    # Only if Raven is installed.
+    if RavenClient is None:
+        return
+
+    if _is_send_to_sentry(exc_info):
+        try:
+            _submit_trace(exc_info)
+        except Exception as exc:
+            # Failed to send - let user know and suggest manual reporting.
+            echo.error(
+                "Encountered an issue while trying to submit the report: {}".format(
+                    str(exc)
+                )
+            )
+            click.echo(_MSG_MANUALLY_REPORT)
+        else:
+            # Sent successfully.
+            click.echo(_MSG_SEND_TO_SENTRY_THANKS)
+
+
+def _print_exception_message(exc_info, debug) -> None:
+    click.echo(_MSG_TRACEBACK_PRINT)
+    echo.error(str(exc_info[1]))
+
+
+def _process_exception(exc_info, debug, trace_filepath):
+    _print_exception_message(exc_info, debug)
+
+    if _is_printable_traceback(exc_info, debug):
+        _print_trace_output(exc_info)
+
+    _handle_sentry_submission(exc_info)
```

Comment:
I was not aware we had a scenario for printing the trace output AND sending to sentry. Should this be an else for the above conditional?

Reply:
I decoupled them with the tweaked ROE. My thought was there may be a case where you do want to report automatically as well as see trace (e.g. SNAPCRAFT_ENABLE_ERROR_REPORTING=y SNAPCRAFT_ENABLE_SILENT_REPORT=y and tty=no).

However, that change in logic does cause a potential issue where tty=no and click.prompt is reached without those variables. So I'll need to fix that regardless.

If you want to maintain the either-or condition, let me know and I'll fix it up.

Reply:
No longer prompt to report if not connected to TTY.  Maintained conditionals as discussed offline.